nexus/src/app/sagas/switch_port_settings_apply.rs

```diff
@@ -284,6 +339,341 @@ async fn spa_undo_ensure_switch_port_settings(
+
+    Ok(())
+}
+async fn spa_undo_ensure_switch_port_bgp_settings(
+    sagactx: NexusActionContext,
+) -> Result<(), Error> {
+    use mg_admin_client::types::DeleteNeighborRequest;
+
+    let osagactx = sagactx.user_data();
+    let nexus = osagactx.nexus();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let settings = sagactx
+        .lookup::<SwitchPortSettingsCombinedResult>("switch_port_settings")
+        .map_err(|e| {
+            ActionError::action_failed(format!(
+                "lookup switch port settings (bgp undo): {e}"
+            ))
+        })?;
+
+    let mg_client: Arc<mg_admin_client::Client> =
+        select_mg_client(&sagactx).await.map_err(|e| {
+            ActionError::action_failed(format!("select mg client (undo): {e}"))
+        })?;
+
+    for peer in settings.bgp_peers {
+        let config = nexus
+            .bgp_config_get(&opctx, peer.bgp_config_id.into())
+            .await
+            .map_err(|e| {
+                ActionError::action_failed(format!("delete bgp config: {e}"))
+            })?;
+
+        mg_client
+            .inner
+            .delete_neighbor(&DeleteNeighborRequest {
+                asn: *config.asn,
+                addr: peer.addr.ip(),
+            })
+            .await
+            .map_err(|e| {
+                ActionError::action_failed(format!("delete neighbor: {e}"))
+            })?;
+    }
+
+    Ok(())
+}
+
+async fn spa_ensure_switch_port_bootstore_network_settings(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let nexus = osagactx.nexus();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let settings = sagactx
+        .lookup::<SwitchPortSettingsCombinedResult>("switch_port_settings")
+        .map_err(|e| {
+            ActionError::action_failed(format!(
+                "lookup switch port settings (bgp undo): {e}"
+            ))
+        })?;
+
+    // Just choosing the sled agent associated with switch0 for no reason.
+    let sa = switch_sled_agent(SwitchLocation::Switch0, &sagactx).await?;
+
+    // Read the current bootstore config, perform the update and write it back.
+    let mut config = read_bootstore_config(&sa).await?;
+    let update = bootstore_update(
+        &nexus,
+        &opctx,
+        params.switch_port_id,
+        &params.switch_port_name,
+        &settings,
+    )
+    .await?;
+    apply_bootstore_update(&mut config, &update)?;
+    write_bootstore_config(&sa, &config).await?;
+
+    Ok(())
+}
+
+async fn spa_undo_ensure_switch_port_bootstore_network_settings(
+    sagactx: NexusActionContext,
+) -> Result<(), Error> {
+    // The overall saga update failed but the bootstore udpate succeeded.
```

Comment:
This is a placeholder until the foundational RPW work @internet-diglett is working on lands. Then all of this will go over to RPW land. This was the best I could come up with outside an RPW.

Reply:
Works for me as we discussed.